enhancements/microshift/advertise-version-details.md

```diff
@@ -0,0 +1,272 @@
+---
+title: advertise-version-details
+authors:
+  - "@dhellmann"
+reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
+  - "@spadgett, console team"
+  - "@wking, upgrades team"
+  - "@sdodson, upgrades team"
+  - "@LalatenduMohanty, upgrades team"
+  - "@soltysh, workloads team"
+approvers:
+  - TBD
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - None
+creation-date: 2022-07-29
+last-updated: 2022-07-29
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/USHIFT-233
+  - https://bugzilla.redhat.com/show_bug.cgi?id=1850656
+---
+
+# Advertise Version Details
+
+## Summary
+
+We have several use cases for allowing authenticated users to discover
+whether they are communicating with OCP or MicroShift, including the
+version of the product. We already have the ClusterVersion resource
+for OCP, but that requires special privileges to read and is not
+present in a MicroShift deployment. This enhancement describes a
+standard that both platforms can use to share platform and version
+information with API callers.
+
+## Motivation
+
+Users configuring applications to deploy on different versions of
+OpenShift may need to adjust their configuration to work differently
+on OCP and MicroShift. Ideally this is not needed, but in practice it
+is useful because of different APIs being available and different
+constraints being in place. It is possible for a cluster admin to read
+the ClusterVersion resource in an OCP cluster to discover that it is
+OCP and what version is running. This approach has two
+drawbacks. First, not all console users have cluster admin privileges,
+which means that the console cannot show them the cluster version
+details. Second, the ClusterVersion API is not present in a MicroShift
+deployment, so instead [we created a
+ConfigMap](https://github.com/openshift/microshift/pull/776) in the
+`openshift-public` namespace containing some basic version
+details. That means a caller must know two different APIs to check the
+version and product they are communicating with. This enhancement
+describes how we can standardize on one to make it easier for all
+authenticated users to obtain basic version details for a cluster.
+
+### User Stories
+
+As a console user, I want to see the version of OCP I am using so that
+I can refer to the correct version of the documentation and API
+references.
```

Comment:
I just learned today that [there's already a knob for this][1] :).  Although it's not clear from the godocs how much of the docs URI goes into `documentationBaseURL`, although [OSD uses `https://docs.openshift.com/dedicated/4/`][2], so version might be included?

[1]: https://github.com/openshift/api/blob/83d1191dd9fc012200ae21fbf682e3b3134ca05f/operator/v1/types_console.go#L96-L101
[2]: https://github.com/openshift/managed-cluster-config/blob/77338848f8f85729fca09fe714e11712b40afe54/deploy/osd-console-branding/osd-branding.console.Patch.yaml#L7

Reply:
Maybe @spadgett can help clarify the use case for the console, if it's indeed different from this.

Reply:
Based on https://bugzilla.redhat.com/show_bug.cgi?id=1850656#c32 it looks like there are different Prometheus queries that the console would need to run for different versions of OCP, so it's not just about docs. I can add that use case as a follow-up patch.

Reply:
I expanded the list of use cases to cover the one for needing an implementation difference that can't be discovered by looking at Kubernetes API versions.